### ISSUE TITLE
createRemoteFileNode with aws s3 format

### ISSUE THREAD

**Issue by cezarneaga** [gatsbyjs/gatsby#6853]
So i am running into some difficulties with `createRemoteFileNode` when used with `urls` coming from AWS S3: the `url` is of format: https://s3.eu-central-1.amazonaws.com/impact-startarium/images/a_cover/original/034e554d-589e-4457-b850-9284d65c2d37. `gatsby-source-filesystem` does its thing and saves the files in .cache as: `034e554d-589e-4457-b850-9284d65c2d37` which results in the fact that when i query for `childImageSharp` => null 
cc: @kylemathews
any ideas how i can overcome this?

**Comment by Chuloo** [comment]
Hi @cezarneaga thanks for bringing this up, can you share a reproduction of the code? This will help us look in and solve problems _if any_. 👍

**Comment by KyleAMathews** [comment]
The problem is that gatsby-source-filesystem doesn't know what type a file is without the file extension. So gatsby-transformer-sharp won't know it can transform it. The fix is the add file extensions.

**Comment by cezarneaga** [comment]
images are created/stored as an object with a property `fileType="image/png"` for example. Could i pass gatsby-transformer-sharp this info somehow?

**Comment by cezarneaga** [comment]
@Chuloo not sure how a code repro would help here. any gatsby starter with this url as image on a post or whatever would produce this. :) 

If really necessary i can of course do it.

**Comment by KyleAMathews** [comment]
You should use that file type to set the extension for the file.

**Comment by KyleAMathews** [comment]
Oh but yeah, you can't do that directly... hmm...

**Comment by cezarneaga** [comment]
yes. :) if i use `url.png` it doesnt resolve anymore

**Comment by KyleAMathews** [comment]
I guess you could transform the newly created node into a node that you set internal.mediaType to image/png

**Comment by KyleAMathews** [comment]
Then gatsby-transformer-sharp would pick that up. This is a bit tricky though.

**Comment by cezarneaga** [comment]
let me give it a try. i guess i should do a check for this type of node in my `createNodesFromEntities` and add it there :)

**Comment by cezarneaga** [comment]
needs to come in here: https://github.com/gatsbyjs/gatsby/blob/1fb19f9ad16618acdac7eda33d295d8ceba7f393/packages/gatsby-source-filesystem/src/create-remote-file-node.js#L224, right @KyleAMathews ?

**Comment by cezarneaga** [comment]
was overdoing it there. i managed to insert it but still `childImageSharp` is `null`. 
<img width="416" alt="screen shot 2018-07-31 at 13 32 07" src="https://user-images.githubusercontent.com/645227/43457208-5eba99ee-94c6-11e8-89d3-53e5dee9ae41.png">

here's the function that downloads the files:
```javascript
export function downloadThumbnails({
  items,
  store,
  cache,
  createNode,
  createNodeId,
}) {
  return Promise.all(
    items.map(async item => {
      let fileNode
      if (item.thumbnail && item.thumbnail.url) {
        try {
          fileNode = await createRemoteFileNode({
            url: item.thumbnail.url,
            store,
            cache,
            createNode,
            createNodeId,
          })
          // here i am adding the correct mediaType
          fileNode.internal.mediaType = item.thumbnail.fileType
        } catch (error) {
          // noop
        }
      }
      if (fileNode) {
        item.image___NODE = fileNode.id
      }
      return item
    })
  )
}
```
i have a feeling though that i need to get `gatsby-transformer-sharp` to accept the lack of extension.
could you please point me to the right location?

thanks,
C

**Comment by mariolopjr** [comment]
I've also hit this snag as well. When using a source plugin that pulls from an API, if the API's files are NOT stored on S3, works perfect. If it is...does not!

However, interestingly, my URLs have file extensions (I am not using AWS per-se, but a clone of AWS -- Digital Ocean Spaces).

I can provide more information if needed or even the API graphql endpoint + GatsbyJS frontend code. It's really interesting that createRemoteFileNode doesn't like the S3 urls (even if it has an extension).

Thoughts?

**Comment by KyleAMathews** [comment]
@mariolopjr can you provide more details about how it doesn't work with URLs that have extensions?

@cezarneaga gatsby-transformer-sharp only knows how to look at extensions atm — it'd be a great change actually to have it use mediaType instead or in addition https://github.com/gatsbyjs/gatsby/blob/a57f7859e4900bdb8585e2e191e09ed59e8d18ea/packages/gatsby-transformer-sharp/src/on-node-create.js#L6

**Comment by cezarneaga** [comment]
@KyleAMathews when you say great change you mean it would be nice to have or its to big to consider the effort? :)
will give it a try.

**Comment by KyleAMathews** [comment]
Nice to have :-)

**Comment by cezarneaga** [comment]
it seems that if the check is performed for empty string too
```const extensions = [`jpeg`, `jpg`, `png`, `webp`, `tif`, `tiff`, ``] ```

it just works. :)

**Comment by cezarneaga** [comment]
@mariolopjr are the files in DO spaces public? if you have permissions set for a specific domain only, sharp will fail.

**Comment by mariolopjr** [comment]
@cezarneaga Yup, they were public. First thing I checked (make sure it wasn't a permissions issues). I swapped over to AWS, though regardless, I think the source plugin exposes the files with extension? Maybe? I noticed extensions on AWS as well.

Additionally, @cezarneaga, will you be taking a stab at the code? Otherwise, I wouldn't mind trying my hand at it. :)

@KyleAMathews Let me re-verify everything with AWS as the backing cloud to the source plugin. I was overriding the endpoint in the source plugin to be able to use DO Spaces. I'll get back to you. :)

**Comment by cezarneaga** [comment]
@mariolopjr i solved it for my case as i mentioned above. See my reply to Kyle above. Not sure if that change impacts other stuff though.

**Comment by mariolopjr** [comment]
@cezarneaga I saw, thanks! You added the empty string to the extensions list. Any repercussions? What if a PDF come through as empty string? I will do some testing later today, I am curious. :)
